lib/spack/spack/fetch_strategy.py

```diff
@@ -792,11 +802,460 @@ def __str__(self):
         return "[go] %s" % self.url
 
 
+class GitRef(object):
+    """A wrapper that knows how to format git refspecs in multiple contexts."""
+    _known_types = ['branch', 'commit', 'tag']  # type: ClassVar[List[str]]
+    _ref_type = None                            # type: str
+    _ref = None                                 # type: str
+
+    @classmethod
+    def known_types(cls):
+        # type: () -> List[str]
+        return cls._known_types
+
+    @property
+    def ref_type(self):
+        # type: () -> str
+        return self._ref_type
+
+    def __init__(self, ref_type, ref):
+        # type: (str, str) -> None
+        if ref_type not in self.known_types():
+            # This is made a TypeError and not an InvalidGitRef because it cannot arise
+            # from invalid user input.
+            raise TypeError('GitRef can only have the types {0}, but was given {1!r}'
+                            .format(self.known_types(), ref_type))
+        self._ref_type = ref_type
+        if not isinstance(ref, six.string_types):
+            # This case can arise from user error when editing package.py files, so we
+            # want to catch this and handle it specially to provide more context.
+            raise InvalidGitRef('git reference was not a string: {0!r}'.format(ref))
+        self._ref = ref
+
+    def __repr__(self):
+        return '{0}(ref_type={1!r}, ref={2!r})'.format(type(self).__name__,
+                                                       self.ref_type, self._ref)
+
+    def __str__(self):
+        return '<{0}>'.format(self.repo_info_for_reference())
+
+    # Matches any 7-40 character hexadecimal string.
+    _commit_rx = re.compile(r'^[0-9a-f]{7,40}$', flags=re.IGNORECASE)
+
+    @classmethod
+    def commit(cls, ref):
+        # type: (str) -> GitRef
+        if not isinstance(ref, six.string_types):
+            # The exception message from `re.match()` does not print the value when
+            # a non-string argument is provided, so we special-case that check here.
+            raise InvalidGitRef('git reference was not a string: {0}'.format(ref))
```

Comment:
yeah and I'm thinking we could even organize all these git regexes / exceptions into a common area for other modules to use (at some future point).

Reply:
I ended up consuming `InvalidGitRef` from within `version.py` after rebasing over your changes in `lookup_commit()` -- let me know if that makes sense to you!